[WebAudio] Implement AudioContext::outputLatency

jyavenard opened the pull request:
#### 420f89a0d4b8256194c08f5e6f7ab3e9288e61ae
<pre>
[WebAudio] Implement AudioContext::outputLatency
<a href="https://bugs.webkit.org/show_bug.cgi?id=285826">https://bugs.webkit.org/show_bug.cgi?id=285826</a>
<a href="https://rdar.apple.com/142794341">rdar://142794341</a>

Reviewed by Youenn Fablet.

Add implementation for Cocoa platforms, other platforms will return 0 for now.

* LayoutTests/imported/w3c/web-platform-tests/webaudio/idlharness.https.window-expected.txt:
* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::AudioContext::outputLatency):
* Source/WebCore/Modules/webaudio/AudioContext.h:
* Source/WebCore/Modules/webaudio/AudioContext.idl:
* Source/WebCore/Modules/webaudio/DefaultAudioDestinationNode.cpp:
(WebCore::DefaultAudioDestinationNode::outputLatency const):
* Source/WebCore/Modules/webaudio/DefaultAudioDestinationNode.h:
* Source/WebCore/platform/audio/AudioDestination.h:
(WebCore::AudioDestination::outputLatency const):
* Source/WebCore/platform/audio/AudioSession.h:
* Source/WebCore/platform/audio/SharedAudioDestination.cpp:
(WebCore::SharedAudioDestinationAdapter::outputLatency const):
(WebCore::SharedAudioDestinationAdapter::protectedDestination const):
(WebCore::SharedAudioDestination::outputLatency const):
(WebCore::SharedAudioDestination::protectedOutputAdapter const):
(WebCore::SharedAudioDestinationAdapter::protectedDestination): Deleted.
(WebCore::SharedAudioDestination::protectedOutputAdapter): Deleted.
* Source/WebCore/platform/audio/SharedAudioDestination.h:
* Source/WebCore/platform/audio/cocoa/AudioDestinationCocoa.cpp:
(WebCore::AudioDestinationCocoa::outputLatency const):
* Source/WebCore/platform/audio/cocoa/AudioDestinationCocoa.h:
* Source/WebCore/platform/audio/cocoa/AudioOutputUnitAdaptor.cpp:
(WebCore::AudioOutputUnitAdaptor::outputLatency const):
* Source/WebCore/platform/audio/cocoa/AudioOutputUnitAdaptor.h:
* Source/WebCore/platform/audio/ios/AudioSessionIOS.h:
* Source/WebCore/platform/audio/ios/AudioSessionIOS.mm:
(WebCore::AudioSessionIOS::outputLatency const):
* Source/WebCore/platform/audio/mac/AudioSessionMac.h:
* Source/WebCore/platform/audio/mac/AudioSessionMac.mm:
(WebCore::AudioSessionMac::outputLatency const):
* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.cpp:
(WebKit::RemoteAudioDestinationManager::createAudioDestination):
(WebKit::RemoteAudioDestinationManager::startAudioDestination):
* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.h:
* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.messages.in:
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.cpp:
(WebKit::RemoteAudioSessionProxy::configuration):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
(WebKit::RemoteAudioDestinationProxy::connection):
(WebKit::RemoteAudioDestinationProxy::startRendering):
(WebKit::RemoteAudioDestinationProxy::outputLatency const):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSessionConfiguration.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSessionConfiguration.serialization.in:

Canonical link: <a href="https://commits.webkit.org/288904@main">https://commits.webkit.org/288904@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a575fca536bfff9268bb62f4c7fb95721b2aebe8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84739 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4464 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39127 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89878 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/35791 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86824 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4553 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12439 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/65943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/35791 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87784 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/3453 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77009 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/46216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3332 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31228 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34865 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/74183 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32034 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91254 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12078 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/8818 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/74422 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12305 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72820 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73548 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18193 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17924 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/16367 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/3526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12030 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/17470 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11864 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15358 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/13610 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/Modules/webaudio/AudioContext.cpp:
```diff
@@ -187,6 +189,20 @@ double AudioContext::baseLatency()
+{
+    lazyInitialize();
+
+    if (noiseInjectionPolicies().contains(NoiseInjectionPolicy::Minimal))
```
I think you could go down this codepath in both cases (`Minimal` and `Enhanced`)

```suggestion
    if (noiseInjectionPolicies())
```

youennf commented on Source/WebCore/Modules/webaudio/AudioContext.cpp:
```diff
@@ -187,6 +189,20 @@ double AudioContext::baseLatency()
+
+    return (MediaTime { static_cast<int64_t>(destination().outputLatency()), static_cast<uint32_t>(sampleRate()) }
+#if USE(AUDIO_SESSION)
+            + MediaTime { static_cast<int64_t>(AudioSession::protectedSharedSession()->outputLatency()), static_cast<uint32_t>(AudioSession::protectedSharedSession()->sampleRate()) }
```
Would it be better to leave destination do this computation?
That way, we can do the computation in GPUProcess and remove the need to send to WebProcess the audio session latency.

jyavenard replied:
Doing so would require to have a way for the RemoteAudioDestination/RemoteAudioDestinationProxy be notified when the AudioSession changes. It's a much bigger change.

And currently the RemoteAudioSession in the WebContent is being notified of changes of AudioSession in the GPU process. So I re-used that.

it made the change much smaller.

I could move that arithmetic in the AudioDestination, but then it kind of blur which latency are we talking about. The AudioDestination itself, the AudioSession, both?

youennf replied:
I would prefer we would move the computation to the AudioDestination.
It is probably the audio destination that will handle setSinkId computation as well, so that seems better this way.

youennf commented on Source/WebCore/platform/audio/mac/AudioSessionMac.mm:
```diff
@@ -483,6 +483,33 @@ static float defaultDeviceTransportIsBluetooth()
     }
 }
 
+size_t AudioSessionMac::outputLatency() const
```
I would tend to move this computation to AudioOuptutUnitAdaptor.
The reason is that AudioContext can be set to a particular device via setSinkId.

To handle the case of audio context playing on the device device and default device is changing, AudioOuptutUnitAdaptor would ideally be notified that the default device is changing and would do the recomputation if it is not set to a particular device via setSinkId.
And then send the updated value to its AudioOuptutUnitAdaptor client.

jyavenard replied:
We don't currently support setSinkId on the AudioContext so I will take this as a follow-up.

jyavenard replied:
I also don't see how we can retrieve the device's latency from the AudioUnit

youennf commented on Source/WebCore/Modules/webaudio/AudioContext.cpp:
```diff
@@ -187,6 +189,20 @@ double AudioContext::baseLatency()
     return static_cast<double>(destination().framesPerBuffer()) / sampleRate();
 }
 
+double AudioContext::outputLatency()
```
It seems odd to return the audiosession latency even if AudioContext is not playing.
I'll tend to move the logic to add the latencies in destination.

jyavenard replied:
Limited it to when the context is playing

youennf reviewed (approved):
r=me.
I would tend to have AudioDestination be the one to do the latency computation.
This eases handling the case of not playing context as well as future setSinkId case.

jyavenard committed:
[WebAudio] Implement AudioContext::outputLatency
https://bugs.webkit.org/show_bug.cgi?id=285826
rdar://142794341

Reviewed by Youenn Fablet.

Add implementation for Cocoa platforms, other platforms will return 0 for now.

* LayoutTests/imported/w3c/web-platform-tests/webaudio/idlharness.https.window-expected.txt:
* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::AudioContext::outputLatency):
* Source/WebCore/Modules/webaudio/AudioContext.h:
* Source/WebCore/Modules/webaudio/AudioContext.idl:
* Source/WebCore/Modules/webaudio/DefaultAudioDestinationNode.cpp:
(WebCore::DefaultAudioDestinationNode::outputLatency const):
* Source/WebCore/Modules/webaudio/DefaultAudioDestinationNode.h:
* Source/WebCore/platform/audio/AudioDestination.h:
(WebCore::AudioDestination::outputLatency const):
* Source/WebCore/platform/audio/AudioSession.h:
* Source/WebCore/platform/audio/SharedAudioDestination.cpp:
(WebCore::SharedAudioDestinationAdapter::outputLatency const):
(WebCore::SharedAudioDestinationAdapter::protectedDestination const):
(WebCore::SharedAudioDestination::outputLatency const):
(WebCore::SharedAudioDestination::protectedOutputAdapter const):
(WebCore::SharedAudioDestinationAdapter::protectedDestination): Deleted.
(WebCore::SharedAudioDestination::protectedOutputAdapter): Deleted.
* Source/WebCore/platform/audio/SharedAudioDestination.h:
* Source/WebCore/platform/audio/cocoa/AudioDestinationCocoa.cpp:
(WebCore::AudioDestinationCocoa::outputLatency const):
* Source/WebCore/platform/audio/cocoa/AudioDestinationCocoa.h:
* Source/WebCore/platform/audio/cocoa/AudioOutputUnitAdaptor.cpp:
(WebCore::AudioOutputUnitAdaptor::outputLatency const):
* Source/WebCore/platform/audio/cocoa/AudioOutputUnitAdaptor.h:
* Source/WebCore/platform/audio/ios/AudioSessionIOS.h:
* Source/WebCore/platform/audio/ios/AudioSessionIOS.mm:
(WebCore::AudioSessionIOS::outputLatency const):
* Source/WebCore/platform/audio/mac/AudioSessionMac.h:
* Source/WebCore/platform/audio/mac/AudioSessionMac.mm:
(WebCore::AudioSessionMac::outputLatency const):
* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.cpp:
(WebKit::RemoteAudioDestinationManager::createAudioDestination):
(WebKit::RemoteAudioDestinationManager::startAudioDestination):
* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.h:
* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.messages.in:
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.cpp:
(WebKit::RemoteAudioSessionProxy::configuration):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
(WebKit::RemoteAudioDestinationProxy::connection):
(WebKit::RemoteAudioDestinationProxy::startRendering):
(WebKit::RemoteAudioDestinationProxy::outputLatency const):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSessionConfiguration.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSessionConfiguration.serialization.in:

Canonical link: https://commits.webkit.org/288904@main
